### DISCUSSION TITLE
Failed to connect... quando provo a configurare MQTT da Core verso il broker

### DISCUSSION THREAD

**Discussion by Domopanda** [indomus/forum#3287] [HUB: Home Assistant]
...rieccomi al punto di ieri con l'istallazione di MQTT che al primo settaggio su Core, dopo aver istallato il broker via composer e apparentemente attivo e funzionante (almeno via portainer mi dice che è avviato ed ok), mi restituisce errore : Failed to connect...
Mi è venuto in mente che ci potrebbe essere un errore di percorso da qualche parte e questo perchè quando ho creato il container la prima volta seguendo la guida, mi son olvidato che io metto tutte le mie cosine in opt, per cui ho lanciato il composer con i percorsi originali da guida, ma quando ho realizzato che non sarebbe stato dove avrei voluto...(stupidamente forse) ho cancellato il container da Portainer, ho aggiornato il percorso nel docker compose.yaml ed ho rilaciato il container e... tutto sembra funzionare.
Di fatto però ho verificato che, seppur il broker sembra attivo e nello stack opt nella dshboard di portainer, in realtà nei percorsi da me sostituiti....non c'è un bel nulla....
questo è nel docker compose:
```  mosquitto:
    container_name: mosquitto
    image: eclipse-mosquitto
    volumes:
    - "/srv/mosquitto/config:/opt/mosquitto/config"
    - "/srv/mosquitto/log:/opt/mosquitto/log"
    - "/srv/mosquitto/data/:/opt/mosquitto/data"
    restart: always
    ports:
    - "1883:1883"  
```

Che dite, potrebbe essere quello l'errore ? Grazie in anticipo...

Nel log di Core, ieri mi dava questo : 
```  ERROR (Thread-6 (_thread_main)) [paho.mqtt.client] failed to receive on   socket: [Errno 104] Connection reset by peer ```
facendo ricerche su github, questo errore potrebbe essere imputato ad un problema di certificato...
Ma io continuo a pensare che ci sia un problema anche di percorsi perchè in realtà non c'è nessuna cartella mosquitto in opt/...
questo il contenuto del mosquitto.config cmq...

``` persistence true
persistence_location /opt/mosquitto/data/
log_dest file /opt/mosquitto/log/mosquitto.log
log_dest stdout
listener 1883
password_file /opt/mosquitto/config/mosquitto.passwd
allow_anonymous false ```

**Comment by snakuzzo** [comment]
Quando fai il bind dei volumi, la parte a sinistra, nel tuo caso `/srv/mosquitto/...`, è quella dell'host, mentre `/opt/mosquito/...` è quella del container.
Tu dici che sotto `/opt` non c'è nulla, ma questa verifica devi farla "DENTRO" il container mosquitto, non sull'host.

Quindi assicurati prima "bene" di questa cosa

**Reply by Domopanda** [reply]
Buongiorno Diego, innanzitutto grazie per la risposta...non sapevo del forum e probabilmente questo è il luogo più adatto per tutte le rogne che Core mi stà regalando :)
In realtà quello che tu mi dici è quello che io pensavo, ma non sono certo di sapere esattamente come fare tale verifica...e per mia ignoranza ovviamente.
Mosquitto sembra si carichi perfettamente questo è il log
```
1731861567: mosquitto version 2.0.20 starting
1731861567: Config loaded from /mosquitto/config/mosquitto.conf.
1731861567: Starting in local only mode. Connections will only be possible from clients running on this machine.
1731861567: Create a configuration file which defines a listener to allow remote access.
1731861567: For more details see https://mosquitto.org/documentation/authentication-methods/
1731861567: Opening ipv4 listen socket on port 1883.
1731861567: Opening ipv6 listen socket on port 1883.
1731861567: mosquitto version 2.0.20 running
1731862054: mosquitto version 2.0.20 terminating
1731862054: mosquitto version 2.0.20 starting
1731862054: Config loaded from /mosquitto/config/mosquitto.conf.
1731862054: Starting in local only mode. Connections will only be possible from clients running on this machine.
1731862054: Create a configuration file which defines a listener to allow remote access.
1731862054: For more details see https://mosquitto.org/documentation/authentication-methods/
1731862054: Opening ipv4 listen socket on port 1883.
1731862054: Opening ipv6 listen socket on port 1883.
1731862054: mosquitto version 2.0.20 running
1732001125: mosquitto version 2.0.20 starting
1732001125: Config loaded from /mosquitto/config/mosquitto.conf.
1732001125: Starting in local only mode. Connections will only be possible from clients running on this machine.
1732001125: Create a configuration file which defines a listener to allow remote access.
1732001125: For more details see https://mosquitto.org/documentation/authentication-methods/
1732001125: Opening ipv4 listen socket on port 1883.
1732001125: Opening ipv6 listen socket on port 1883.
1732001125: mosquitto version 2.0.20 running
 ```
 Invece questo è un estratto dell'inspect...fino ai binding 
 ```
AppArmorProfile
Args [ /usr/sbin/mosquitto, -c, /mosquitto/config/mosquitto.conf ]
Config
AttachStderr true
AttachStdin false
AttachStdout true
Cmd [ /usr/sbin/mosquitto, -c, /mosquitto/config/mosquitto.conf ]
Domainname
Entrypoint [ /docker-entrypoint.sh ]
Env [ PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin, VERSION=2.0.20, DOWNLOAD_SHA256=ebd07d89d2a446a7f74100ad51272e4a8bf300b61634a7812e19f068f2759de8, GPG_KEYS=A0D6EEA1DCAE49A635A3B2F0779B22DFB3E717B7, LWS_VERSION=4.2.1, LWS_SHA256=842da21f73ccba2be59e680de10a8cce7928313048750eb6ad73b6fa50763c51 ]
ExposedPorts { 1883/tcp: [object Object] }
Hostname f640c9ef7b19
Image eclipse-mosquitto
Labels { com.docker.compose.config-hash: 0fa51f6482e9035f6749dff156615cda8fe7d232efefb8dcbcf5933b08707311, com.docker.compose.container-number: 1, com.docker.compose.depends_on: , com.docker.compose.image: sha256:4372cc18130828529587cde9f1b17628e192c805c0ab21210f19b5492642f9f7, com.docker.compose.oneoff: False, com.docker.compose.project: opt, com.docker.compose.project.config_files: /opt/docker-compose.yaml, com.docker.compose.project.working_dir: /opt, com.docker.compose.service: mosquitto, com.docker.compose.version: 2.29.7, description: Eclipse Mosquitto MQTT Broker, maintainer: Roger Light <roger@atchoo.org> }
OnBuild
OpenStdin false
StdinOnce false
Tty false
User
Volumes
/mosquitto/data { }
/mosquitto/log { }
WorkingDir /
Created 2024-11-17T16:39:26.615083617Z
Driver overlay2
ExecIDs
GraphDriver { Data: [object Object], Name: overlay2 }
HostConfig
AutoRemove false
Binds
0 /srv/mosquitto/log:/opt/mosquitto/log:rw
1 /srv/mosquitto/data/:/opt/mosquitto/data:rw
2 /srv/mosquitto/config:/opt/mosquitto/config:rw
 ```
Mi permetto di postarli perchè non capisco bene cosa dovrei verificare esattamente, tuttavia se quello che ho postato è inutile, cortesemente solo suggeriscimi più specificatamente cosa devo verificare e lo faccio volentieri...se ti và :)
Grazie in anticipo ....come sempre

**Reply by snakuzzo** [reply]
Allora...
per verificare i file dentro il container, ti basta aprire una shell del container in questo modo:

`docker exec -it <nomecontainer> sh`

In ogni caso, hai sbagliato a fare Il bind dei volumi.

tu hai fatto...

```yaml
volumes:
  - "/srv/mosquitto/config:/opt/mosquitto/config"
  - "/srv/mosquitto/log:/opt/mosquitto/log"
  - "/srv/mosquitto/data/:/opt/mosquitto/data"
```

ma lui, come puoi vedere anche facilmente dal log che hai postato, si aspetta la configurazione sotto `/mosquitto/config/`

es.

```
1731861567: Config loaded from /mosquitto/config/mosquitto.conf.
```

In sostanza, detta in maniera barbara, tu stai mettendo i file in una directory che lui non si caga minimamente, per cui non andando in override sui file di default, lui prende appunto i file di configurazione di default.

Ora, ammesso che sul tuo host i file siano stati correttamente creati nei path "a sinistra", il bind che devi fare è il seguente (quello dei log non ti occorre)...

```yaml
volumes:
  - "/srv/mosquitto/config:/mosquitto/config"
  - "/srv/mosquitto/data/:mosquitto/data"
```

**Reply by Domopanda** [reply]
Grazie molte Diego, era esattamente quello che pensavo ma non ne ero sicuro, infatti...come avevo detto all'inizio, dopo la prima istallazione evidentemente corretta del continer del broker, sono stato io che l'ho cancellata per farla ricreare nella dir opt e probabilmente è da li che è cominciato il casino.
Procedendo come da te suggerito ed a scanso di equivoci, ho rimosso il container di mosquitto da portainer, ho modificato il docker compose cosi con i percorsi corretti:
```
  mosquitto:
    container_name: mosquitto
    image: eclipse-mosquitto
    volumes:
    - "/srv/mosquitto/config:/mosquitto/config"
    - "/srv/mosquitto/log:/mosquitto/log"
    - "/srv/mosquitto/data/:/mosquitto/data"
    restart: always
    ports:
    - "1883:1883"
 ```
 ho quindi rilanciato ``` docker compose up -d mosquitto  ``` ricreando il broker

e mi sono accertato che i il file moquitto.conf fosse corretto con i nuovi percorsi

 ```
 persistence true
persistence_location /mosquitto/data/
log_dest file /mosquitto/log/mosquitto.log
log_dest stdout
listener 1883
password_file /mosquitto/config/mosquitto.passwd
allow_anonymous false

 ```
 A questo punto, in buonafede, mi sembrava che tutto fosse corretto ma....porca pupazza, mi restituisce sempre lo stesso errore...
 nel frattempo sono due ore che provo variabili al tema, ma forse ho semplicemente incasinato tutto io e non sò come uscirne, eppure ero certo fosse una cazzata...
